src/main/java/cbot/io/Parser.java

```diff
@@ -0,0 +1,362 @@
+package cbot.io;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+
+import cbot.task.Deadline;
+import cbot.task.Event;
+import cbot.task.Task;
+import cbot.task.TaskList;
+import cbot.time.TimeStuff;
+
+/**
+ * Handles much of the parsing of user inputs.
+ */
+public class Parser {
+    private Command c;
+    private String text;
+
+    /**
+     * Constructs a new instance to process the user's current input. Attempts to recognize the Command,
+     * and the rest of the input body.
+     *
+     * @param input The user's input.
+     * @throws PoorInputException If the input command is not recognized.
+     * @see Command
+     */
+    public Parser(String input) throws PoorInputException {
+        this.c = null;
+        this.text = "";
+
+        boolean matchFound = false;
+
+        if (input.contains(Task.SEP)) {
+            throw new BadInputException("Please avoid using: \"" + Task.SEP + "\"");
+        }
+
+        for (Command c : Command.values()) {
+            if (c.matches(input)) {
+                this.c = c;
+                this.text = c.getText(input);
+                matchFound = true;
+                break;
+            }
+        }
+
+        if (!matchFound) {
+            throw new PoorInputException("Sorry, I don't recognize that command :<");
+        }
+    }
+
+    /**
+     * Returns true if the stored Command is the BYE Command.
+     *
+     * @return true if the stored Command is BYE.
+     */
+    public boolean isBye() {
+        return (this.c == Command.BYE);
+    }
+
+    /**
+     * Returns true if the stored Command requires the task list to be saved.
+     *
+     * @return Whether the task list needs to be saved after the Command.
+     */
+    public boolean needSave() {
+        return this.c.needSave();
+    }
+
+    /**
+     * Processes the Command and the rest of the input. If needed, the TaskList is modified and the UI is given
+     * outputs to print. The main bulk of Cbot's parsing is done here.
+     *
+     * @param tl The current list of tasks.
+     * @return The String response to the command.
+     * @throws PoorInputException If the input text is improper or erroneous.
+     * @throws DateTimeParseException If some provided datetime is not in a recognized format.
+     */
+    public String respond(TaskList tl) throws PoorInputException, DateTimeParseException {
```

Comment:
This huge method can perhaps be broken down into smaller functions. It's quite hard to follow the logic of this very long method and you have to indent quite heavily to implement all the logic.

Reply:
Hm, I considered turning each Command into its own Class, and have the functionality be within those, but that felt like too much work hah
Thank you for pointing this out! I'll definitely think about what I can do to make it more readable.